File: 백준/Platinum/2188. 축사 배정/축사 배정.c
#include<stdio.h>
#include<malloc.h>
#include<stdbool.h>
#include<memory.h>

int *S=NULL, **cow=NULL, *cow_count=NULL;
bool *visited=NULL;

bool dfs(int current)
{
	for(int i=0;i<cow_count[current];++i)
	{
		int next=cow[current][i];

		if(!visited[next])
		{
			visited[next]=true;

			if(!S[next] || dfs(S[next]))
			{
				S[next]=current;
				return true;
			}
		}
	}

	return false;
}

int main(void)
{
	int N, M, count=0;

	scanf("%d%d", &N, &M);
	cow=(int **)malloc((N+1)*sizeof(int *));
	cow_count=(int *)malloc((N+1)*sizeof(int));
	S=(int *)calloc(M+1,sizeof(int));
	visited=(bool *)malloc((M+1)*sizeof(bool));

	for(int n=1;n<=N;++n)
	{
		scanf("%d", &cow_count[n]);
		cow[n]=(int *)malloc(cow_count[n]*sizeof(int));

		for(int i=0;i<cow_count[n];++i)
			scanf("%d", &cow[n][i]);
	}

	for(int n=1;n<=N;++n)
	{
		memset(visited+1,0,M);
		count+=dfs(n);
	}

	printf("%d", count);
	while(N)
		free(cow[N--]);
	free(cow);
	free(cow_count);
	free(S);
	free(visited);
	return 0;
}

File: 백준/Platinum/2188. 축사 배정/축사 배정.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

    static int[] S;
    static List<Integer>[] cow;
    static boolean[] visited;

    static boolean dfs(int current) {
        for (Integer next : cow[current]) {
            if (!visited[next]) {
                visited[next] = true;

                if (S[next] == 0 || dfs(S[next])) {
                    S[next] = current;
                    return true;
                }
            }
        }

        return false;
    }

    public static void main(String[] args) throws IOException {
        int count;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        cow = new List[N + 1];
        S = new int[M + 1];
        visited = new boolean[M + 1];

        for (int n = 1; n <= N; ++n) {
            st = new StringTokenizer(br.readLine());

            count = Integer.parseInt(st.nextToken());
            cow[n] = new ArrayList<>();

            while (count-- > 0) {
                cow[n].add(Integer.parseInt(st.nextToken()));
            }
        }

        for (int n = 1; n <= N; ++n) {
            Arrays.fill(visited, false);
            dfs(n);
        }

        count = 0;
        while (M > 0) {
            if (S[M--] > 0) {
                ++count;
            }
        }

        System.out.println(count);
    }
}